File: library/src/main/java/org/fbme/lib/iec61499/descriptors/FBTypeDescriptorUtils.java
package org.fbme.lib.iec61499.descriptors;

import org.fbme.lib.iec61499.declarations.*;
import org.fbme.lib.iec61499.fbnetwork.EntryKind;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

class FBTypeDescriptorUtils {

    private FBTypeDescriptorUtils() {
    }

    public static @NotNull List<FBPortDescriptor> getEventInputPorts(@NotNull FBInterfaceDeclaration declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<EventDeclaration> inputEvents = declaration.getInputEvents();
        for (int i = 0; i < inputEvents.size(); i++) {
            EventDeclaration eventDeclaration = inputEvents.get(i);
            res.add(new FBPortDescriptor(eventDeclaration.getName(), EntryKind.EVENT, i, true, true));
        }
        return res;
    }

    public static @NotNull List<FBPortDescriptor> getEventOutputPorts(@NotNull FBInterfaceDeclaration declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<EventDeclaration> outputEvents = declaration.getOutputEvents();
        for (int i = 0; i < outputEvents.size(); i++) {
            EventDeclaration eventDeclaration = outputEvents.get(i);
            res.add(new FBPortDescriptor(eventDeclaration.getName(), EntryKind.EVENT, i, false, true));
        }
        return res;
    }

    public static @NotNull List<FBPortDescriptor> getDataInputPorts(@NotNull FBInterfaceDeclaration declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<ParameterDeclaration> inputParameters = declaration.getInputParameters();
        for (int i = 0; i < inputParameters.size(); i++) {
            ParameterDeclaration parameterDeclaration = inputParameters.get(i);
            res.add(new FBPortDescriptor(parameterDeclaration.getName(), EntryKind.DATA, i, true, true));
        }
        return res;
    }


    public static @NotNull List<FBPortDescriptor> getDataOutputPorts(@NotNull FBInterfaceDeclaration declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<ParameterDeclaration> outputParameters = declaration.getOutputParameters();
        for (int i = 0; i < outputParameters.size(); i++) {
            ParameterDeclaration parameterDeclaration = outputParameters.get(i);
            res.add(new FBPortDescriptor(parameterDeclaration.getName(), EntryKind.DATA, i, false, true));
        }
        return res;
    }

    public static @NotNull List<FBPortDescriptor> getSocketPorts(@NotNull FBInterfaceDeclarationWithAdapters declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<PlugDeclaration> outputParameters = declaration.getPlugs();
        for (int i = 0; i < outputParameters.size(); i++) {
            PlugDeclaration plugDeclaration = outputParameters.get(i);
            res.add(new FBPortDescriptor(plugDeclaration.getName(), EntryKind.ADAPTER, i, true, true));
        }
        return res;
    }

    public static @NotNull List<FBPortDescriptor> getPlugPorts(@NotNull FBInterfaceDeclarationWithAdapters declaration) {
        List<FBPortDescriptor> res = new ArrayList<>();
        List<SocketDeclaration> outputParameters = declaration.getSockets();
        for (int i = 0; i < outputParameters.size(); i++) {
            SocketDeclaration socketDeclaration = outputParameters.get(i);
            res.add(new FBPortDescriptor(socketDeclaration.getName(), EntryKind.ADAPTER, i, false, true));
        }
        return res;
    }

    public static @NotNull List<Integer> getAssociatedVariablesForInputEvent(@NotNull FBInterfaceDeclaration declaration, int eventNumber) {
        List<ParameterDeclaration> inputParameters = declaration.getInputParameters();
        List<Integer> list = new ArrayList<>();
        for (EventAssociation eventAssociation : declaration.getInputEvents().get(eventNumber).getAssociations()) {
            int parameterIndex = inputParameters.indexOf(eventAssociation.getParameterReference().getTarget());
            list.add(parameterIndex);
        }
        return list;
    }

    public static @NotNull List<Integer> getAssociatedVariablesForOutputEvent(@NotNull FBInterfaceDeclaration declaration, int eventNumber) {
        List<ParameterDeclaration> outputParameters = declaration.getOutputParameters();
        List<Integer> list = new ArrayList<>();
        for (EventAssociation eventAssociation : declaration.getOutputEvents().get(eventNumber).getAssociations()) {
            int parameterIndex = outputParameters.indexOf(eventAssociation.getParameterReference().getTarget());
            list.add(parameterIndex);
        }
        return list;
    }
}

File: code/library/src/main/kotlin/org/fbme/lib/iec61499/descriptors/FBTypeDescriptorUtils.kt
package org.fbme.lib.iec61499.descriptors

import org.fbme.lib.iec61499.declarations.FBInterfaceDeclaration
import org.fbme.lib.iec61499.declarations.FBInterfaceDeclarationWithAdapters
import org.fbme.lib.iec61499.fbnetwork.EntryKind

internal object FBTypeDescriptorUtils {
    @JvmStatic
    fun getEventInputPorts(declaration: FBInterfaceDeclaration): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val inputEvents = declaration.inputEvents
        for (i in inputEvents.indices) {
            val eventDeclaration = inputEvents[i]
            res.add(FBPortDescriptor(eventDeclaration.name, EntryKind.EVENT, i, true, true, eventDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getEventOutputPorts(declaration: FBInterfaceDeclaration): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val outputEvents = declaration.outputEvents
        for (i in outputEvents.indices) {
            val eventDeclaration = outputEvents[i]
            res.add(FBPortDescriptor(eventDeclaration.name, EntryKind.EVENT, i, false, true, eventDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getDataInputPorts(declaration: FBInterfaceDeclaration): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val inputParameters = declaration.inputParameters
        for (i in inputParameters.indices) {
            val parameterDeclaration = inputParameters[i]
            res.add(FBPortDescriptor(parameterDeclaration.name, EntryKind.DATA, i, true, true, parameterDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getDataOutputPorts(declaration: FBInterfaceDeclaration): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val outputParameters = declaration.outputParameters
        for (i in outputParameters.indices) {
            val parameterDeclaration = outputParameters[i]
            res.add(FBPortDescriptor(parameterDeclaration.name, EntryKind.DATA, i, false, true, parameterDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getSocketPorts(declaration: FBInterfaceDeclarationWithAdapters): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val inputParameters = declaration.sockets
        for (i in inputParameters.indices) {
            val socketDeclaration = inputParameters[i]
            res.add(FBPortDescriptor(socketDeclaration.name, EntryKind.ADAPTER, i, true, true, socketDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getPlugPorts(declaration: FBInterfaceDeclarationWithAdapters): List<FBPortDescriptor> {
        val res: MutableList<FBPortDescriptor> = ArrayList()
        val outputParameters = declaration.plugs
        for (i in outputParameters.indices) {
            val plugDeclaration = outputParameters[i]
            res.add(FBPortDescriptor(plugDeclaration.name, EntryKind.ADAPTER, i, false, true, plugDeclaration))
        }
        return res
    }

    @JvmStatic
    fun getAssociatedVariablesForInputEvent(declaration: FBInterfaceDeclaration, eventNumber: Int): List<Int> {
        val inputParameters = declaration.inputParameters
        val list: MutableList<Int> = ArrayList()
        for (eventAssociation in declaration.inputEvents[eventNumber].associations) {
            val parameterIndex = inputParameters.indexOf(eventAssociation.parameterReference.getTarget())
            list.add(parameterIndex)
        }
        return list
    }

    @JvmStatic
    fun getAssociatedVariablesForOutputEvent(declaration: FBInterfaceDeclaration, eventNumber: Int): List<Int> {
        val outputParameters = declaration.outputParameters
        val list: MutableList<Int> = ArrayList()
        for (eventAssociation in declaration.outputEvents[eventNumber].associations) {
            val parameterIndex = outputParameters.indexOf(eventAssociation.parameterReference.getTarget())
            list.add(parameterIndex)
        }
        return list
    }
}
